catalyst-gateway-crates/c509-certificate/src/extensions/mod.rs

```diff
@@ -0,0 +1,176 @@
+///   extension values.
+/// * `critical` - A boolean indicating whether the extension is critical.
+pub struct Extension {
+    pub extension_type: ExtensionRegistry,
+    pub extension_value: ExtensionValue,
+    pub critical: bool,
+}
+
+/// Currently implemented values for an extension.
+#[allow(unused)]
+pub enum ExtensionValue {
+    /// Alternative names for the issuer.
+    IssuerAltName(AltName),
+    /// Alternative names for the subject.
+    SubjectAltName(AltName),
+}
+
+impl CborEncoder for ExtensionValue {
+    /// Encodes a `ExtensionValue` using the provided CBOR encoder.
+    ///
+    /// # Arguments
+    ///
+    /// * `encoder` - A mutable reference to a `minicbor::Encoder` to encode the
+    ///   extensions into.
+    fn encode(&self, encoder: &mut Encoder<&mut Vec<u8>>) {
+        match self {
+            ExtensionValue::IssuerAltName(alt_name) => alt_name.encode(encoder),
+            ExtensionValue::SubjectAltName(alt_name) => alt_name.encode(encoder),
+        }
+    }
+}
+
+#[allow(unused)]
+/// Enum of possible C509 extension types.
+/// Please refer to the [c509-certificate](https://datatracker.ietf.org/doc/draft-ietf-cose-cbor-encoded-cert/09/) Section 9.4 C509 Extensions Registry for more information.
+pub enum ExtensionRegistry {
+    SubjectKeyIdentifier = 1,
+    KeyUsage = 2,
+    SubjectAltName = 3,
+    BasicConstraints = 4,
+    CRLDistributionPoints = 5,
+    CertificatePolicies = 6,
+    AuthorityKeyIdentifier = 7,
+    ExtendedKeyUsage = 8,
+    AuthorityInfoAccess = 9,
+    SignedCertificateTimestampList = 10,
+    SubjectDirectoryAttributes = 24,
+    IssuerAltName = 25,
+    NameConstraints = 26,
+    PolicyMappings = 27,
+    PolicyConstraints = 28,
+    FreshestCRL = 29,
+    InhibitAnyPolicy = 30,
+    SubjectInfoAccess = 31,
+    IPResources = 32,
+    ASResources = 33,
+    IPResourcesBlocksV2 = 34,
+    ASResourcesV2 = 35,
+    BiometricInfo = 36,
+    PrecertificateSigningCert = 37,
+    OCSPNoCheck = 38,
+    QualifiedCertificateStatements = 39,
+    SMIMECapabilities = 40,
+    TLSFeatures = 41,
+    ChallengePassword = 255,
+}
+
+/// Determines if an Extension is critical or not.
+///
+/// # Parameters
+///
+/// * `critical` - A boolean indicating if the extension is critical.
+///
+/// # Returns
+///
+/// Returns `-1` if the Extension is critical, `1` otherwise.
+pub(crate) fn is_critical(critical: bool) -> i16 {
+    if critical {
+        return -1;
+    }
+    1
+}
+
+/// Encodes a list of Extensions into CBOR format.
+/// This function iterates over each `Extension` in the provided `extensions` vector and
+/// encodes
+///
+/// # Parameters
+///
+/// * `extensions` - A vector of `Extension` objects to be encoded.
+/// * `encoder` - A mutable reference to a `minicbor::Encoder` to encode the extensions
+///   into.
+pub(crate) fn encode_extensions(extensions: Extensions, encoder: &mut Encoder<&mut Vec<u8>>) {
```

Comment:
This should be `impl` on the `Extensions` type, rather than a function which takes it as a parameter.
Also `Extension` should have an `encode` function implemented on it, which just encodes a single extension.
This function would then iterate and just call `ext.encode()` and add the result to the list.
This should follow on up, so `ExtensionRegistry` should ALSO have an `encode(critical)` function implemented on it which knows how to encode itself.

Each of these types can be in their own file and self contained from the lowest layers up.